File: AngularJS/src/org/angular2/lang/metadata/MetadataJsonStubBuilder.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.metadata;

import com.intellij.json.psi.impl.JsonFileImpl;
import com.intellij.lang.Language;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiFile;
import com.intellij.psi.stubs.BinaryFileStubBuilder;
import com.intellij.psi.stubs.Stub;
import com.intellij.util.indexing.FileContent;
import org.angular2.lang.metadata.psi.MetadataFileImpl;
import org.angular2.lang.metadata.stubs.MetadataFileStubImpl;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class MetadataJsonStubBuilder implements BinaryFileStubBuilder {

  @Override
  public boolean acceptsFile(@NotNull VirtualFile file) {
    return file.getFileType() == MetadataJsonFileType.INSTANCE;
  }

  @Nullable
  @Override
  public Stub buildStubTree(@NotNull FileContent fileContent) {
    PsiFile psiFile = fileContent.getPsiFile();
    if (!(psiFile instanceof JsonFileImpl)) return null;

    Document document = FileDocumentManager.getInstance().getCachedDocument(fileContent.getFile());
    Project project = fileContent.getProject();
    if (project == null) {
      project = psiFile.getProject();
    }
    if (document != null) {
      PsiFile existingPsi = PsiDocumentManager.getInstance(project).getPsiFile(document);
      if (existingPsi instanceof JsonFileImpl) {
        psiFile = existingPsi;
      }
    }

    JsonFileImpl jsonFile = (JsonFileImpl)psiFile;
    for (Language language: jsonFile.getViewProvider().getLanguages()) {
      if (language instanceof MetadataLanguage) {
        MetadataFileImpl metadataFile = (MetadataFileImpl)jsonFile.getViewProvider().getPsi(language);
        MetadataFileStubImpl result = new MetadataFileStubImpl(metadataFile, ((MetadataLanguage)language).getFileElementType());
        if (jsonFile.getTopLevelValue() != null) {
          ((MetadataLanguage)language).createRootStub(result, jsonFile.getTopLevelValue());
        }
        return result;
      }
    }
    return null;

  }

  @Override
  public int getStubVersion() {
    return 0;
  }
}

File: AngularJS/src/org/angular2/lang/metadata/MetadataJsonStubBuilder.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.lang.metadata

import com.intellij.json.JsonLanguage
import com.intellij.json.psi.impl.JsonFileImpl
import com.intellij.openapi.fileEditor.impl.LoadTextUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiFileFactory
import com.intellij.psi.stubs.BinaryFileStubBuilder
import com.intellij.psi.stubs.Stub
import com.intellij.util.indexing.FileContent
import org.angular2.lang.metadata.stubs.MetadataFileStubImpl

class MetadataJsonStubBuilder : BinaryFileStubBuilder {

  override fun acceptsFile(file: VirtualFile): Boolean {
    return file.fileType is MetadataJsonFileType
  }

  override fun buildStubTree(fileContent: FileContent): Stub {
    val fileType = fileContent.fileType as MetadataJsonFileType

    val text = LoadTextUtil.getTextByBinaryPresentation(
      fileContent.content, fileContent.file)

    val jsonFile = PsiFileFactory
      .getInstance(fileContent.project)
      .createFileFromText(JsonLanguage.INSTANCE, text) as JsonFileImpl

    val result = MetadataFileStubImpl(null, fileType.fileElementType)
    jsonFile.topLevelValue?.let {
      fileType.createRootStub(result, it)
    }
    return result
  }

  override fun getStubVersion(): Int {
    return 22
  }
}
